mbzirc_custom/mbzirc_naive_spinning_radar/src/MaritimeRadar.cc

```diff
@@ -0,0 +1,267 @@
+  this->linePub  = this->node.Advertise<msgs::Float_V>(this->radarSpokeTopic);
+}
+
+///////////////////////////////////////////////////
+void MaritimeRadar::PostUpdate(
+  const ignition::gazebo::UpdateInfo &_info,
+  const ignition::gazebo::EntityComponentManager &_ecm)
+{
+  if (_info.paused)
+    return;
+
+  auto jointPosComp =
+    _ecm.Component<components::JointPosition>(this->jointEntity);
+  if (jointPosComp == nullptr)
+  {
+    ignerr << "No joint position component found for entity ["
+           << this->jointEntity << "]\n";
+    return;
+  }
+  // We just created the joint position component, give one iteration for the
+  // physics system to update its size
+  if (jointPosComp == nullptr || jointPosComp->Data().empty())
+    return;
+
+  // Sanity check: Make sure that the joint has an index
+  if (jointPosComp->Data().size() == 0)
+  {
+    return;
+  }
+
+  std::size_t index = floor(
+    fmod(jointPosComp->Data()[0], 2 * IGN_PI) /  angularResolution);
+
+  std::lock_guard<std::mutex> lock(this->mtx);
+  if (index != this->radarBinIndex)
+  {
+    this->PublishScan();
+    this->ClearScanBuffer((index) % this->radarBin.size());
+    this->radarBinIndex = index;
+    this->numBeams = 0;
+  }
+}
+
+///////////////////////////////////////////////////
+void MaritimeRadar::PreUpdate(
+  const ignition::gazebo::UpdateInfo &_info,
+  ignition::gazebo::EntityComponentManager &_ecm)
+{
+  if (_info.paused)
+    return;
+
+  // We need to track the joint position in order to know when to publish
+  auto jointPosComp =
+    _ecm.Component<components::JointPosition>(this->jointEntity);
+  if (jointPosComp == nullptr)
+  {
+    _ecm.CreateComponent(jointEntity, components::JointPosition());
+  }
+}
+
+///////////////////////////////////////////////////
+void MaritimeRadar::PublishScan()
+{
+  if(this->radarBin.size() < radarBinIndex +1)
+    return;
+  // Publish just a single radar bin
+  //
+  ignition::msgs::Float_V lineBin;
+  // First field of the float array is the angle
+  lineBin.add_data(this->radarBinIndex * this->angularResolution);
+  // Second field is the angular resolution
+  lineBin.add_data(this->angularResolution);
+  // Third field is the linear resolution
+  lineBin.add_data(this->resolution);
+  // Rest of the fields are the range bins
+  for (std::size_t i = 0; i < this->radarBin[radarBinIndex].size(); ++i)
+  {
+    // Convert to DB reflection
+    // Equation can be found here: https://en.wikipedia.org/wiki/Decibel
+    // TODO(arjo): Add some noise to this.
+    // Sea state vs radar noise model can be found here:
+    // https://www.mathworks.com/help/radar/ug/sea-clutter-simulation-for-maritime-radar-system.html
+    lineBin.add_data(
+      10 * log10(this->radarBin[radarBinIndex][i]/this->numBeams));
+  }
+  linePub.Publish(lineBin);
+
+  // The code below this point is for publishing debug images with a polar plot
+  // of radar hits/misses
+  // Convert to a message and publish
+  if(this->radarBinIndex == 0)
```

Comment:
perf optimization: draw debug image only if `this->debugPub` has subscribers

Reply:
https://github.com/osrf/mbzirc/pull/127/commits/52aa818a3355b82bd7b16762183248f8c2b3e5c8